packages/flutter/lib/src/widgets/focus_manager.dart

```diff
@@ -68,330 +512,347 @@ class FocusNode extends ChangeNotifier {
+/// {@macro flutter.widgets.focus_manager.focus.keyEvents}
+///
+/// See also:
+///
+///   * [Focus], a widget that manages a [FocusNode] and provides access to
+///     focus information and actions to its descendant widgets.
+///   * [FocusScope], a widget that manages a [FocusScopeNode] and provides
+///     access to scope information and actions to its descendant widgets.
+///   * [FocusAttachment], a widget that connects a [FocusScopeNode] to the
+///     focus tree.
+///   * [FocusManager], a singleton that manages the focus and distributes key
+///     events to focused nodes.
+class FocusScopeNode extends FocusNode {
+  /// Creates a FocusScope node.
+  ///
+  /// All parameters are optional.
+  FocusScopeNode({
+    String debugLabel,
+    FocusOnKeyCallback onKey,
+  }) : super(debugLabel: debugLabel, onKey: onKey);
+
+  @override
+  FocusScopeNode get nearestScope => this;
 
-  /// Adopts the given scope if it is the first focus of another scope.
+  /// Returns true if this scope is the focused child of its parent scope.
+  bool get isFirstFocus => enclosingScope.focusedChild == this;
+
+  /// Returns the child of this node that should receive focus if this scope
+  /// node receives focus.
   ///
-  /// A widget that sets a scope as the first focus of another scope should
-  /// call this method during its `build` method in case the widget is moved
-  /// from one location in the tree to another location that has a different
-  /// focus scope.
+  /// If [hasFocus] is true, then this points to the child of this node that is
+  /// currently focused.
   ///
-  /// If the given scope is not the first focus of its old parent, the scope
-  /// is simply detached from its old parent.
-  void reparentScopeIfNeeded(FocusScopeNode child) {
-    assert(child != null);
-    if (child._parent == null || child._parent == this)
-      return;
-    if (child.isFirstFocus) {
-      setFirstFocus(child);
+  /// Returns null if there is no currently focused child.
+  FocusNode get focusedChild {
+    assert(_focusedChildren.isEmpty || _focusedChildren.last.enclosingScope == this,
+      'Focused child does not have the same idea of its enclosing scope as the scope does.');
+    return _focusedChildren.isNotEmpty ? _focusedChildren.last : null;
+  }
+
+  // A stack of the children that have been set as the focusedChild, most recent
+  // last (which is the top of the stack).
+  final List<FocusNode> _focusedChildren = <FocusNode>[];
+
+  /// Make the given [scope] the active child scope for this scope.
+  ///
+  /// If the given [scope] is not yet a part of the focus tree, then add it to
+  /// the tree as a child of this scope. If it is already part of the focus
+  /// tree, the given scope must be a descendant of this scope.
+  void setFirstFocus(FocusScopeNode scope) {
+    assert(scope != null);
+    if (scope._parent == null) {
+      _reparent(scope);
+    }
+    assert(scope.ancestors.contains(this));
```

Comment:
Shouldn't this get a big assert message like the other similar asserts?

Reply:
Done.